<a href="https://colab.research.google.com/github/RomaViraj/APT/blob/master/ATV_Lesson_17_confusion%20matrix_v0_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 17: Hunting Exoplanets In Space - Model Evaluation

**WARNING:** The reference notebook is meant **ONLY** for a teacher. Please **DO NOT** share it with any student. 

The contents of the reference notebook are meant only to prepare a teacher for a class. To conduct the class, use the class copy of the reference notebook. The link and the instructions for the same are provided in the **Notes To The Teacher** section.


|Particulars|Description|
|-|-|
|**Topic**|Hunting Exoplanets In Space - Model Evaluation|
|||
|**Class Description**|In this class, a student will learn how to evaluate a classification model based on the precision, recall and f1-score values|
|||
|**Class**|C17|
|||
|**Class Time**|45 minutes|
|||
|**Goal**|Create a confusion matrix|
||Compute the precision value|
||Estimate the recall value|
||Calculate the f1-score value|
|||
|**Teacher Resources**|Google Account|
||Link to Lesson 17 Colab reference notebook|
||Laptop with internet connectivity|
||Earphones with mic|
|||
|**Student Resources**|Google Account|
||Laptop with internet connectivity|
||Earphones with mic|

---

### Class Structure

A class is divided into three parts, as shown in the table below.

|Parts|Duration|
|-|-|
|**Warm-Up**|5 minutes|
|||
|**Teacher-Student Activities**|35 minutes|
|||
|**Wrap-Up**|5 minutes|

---

### Notes To The Teacher

Before beginning the class:

1. Open the **ATV - Lesson 17 - Class Copy - v0.2** file by clicking on the link provided in the **Activities** section under the title **Hunting Exoplanets In Space - Model Evaluation (Class Copy)**.  

2. After opening the file mentioned in the first point, create its duplicate copy by following the steps described below:

    - Click on the **File** menu. A new drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

    - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

    - In the duplicate copy, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/2_share_button.png' width=500>

    - Click on the tiny downward arrow next to **Anyone with the link can view** text. A drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/3_anyone_can_view.png' width=500>

    - Click on the **More...** option. A new page on the dialog box will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/4_more_option.png' width=500>

    - Click on the circle next to **On - Anyone with the link** option. Now, go down and click on the tiny downward box next to **Can edit** text. A drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/5_anyone_with_the_link.png' width=500>

    - Click on the **Can edit** option. Then click on the **Save** button. You will be directed back to the first page of the dialog box.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/6_can_edit.png' width=500>

    - Make sure that under the **Link sharing on** section, **Anyone with the link can edit** option is selected. 

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/7_anyone_can_edit.png' width=500>

    - Then click on the **Done** button.

3. After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Lesson17** format. 

4. Now, finish the **Warm-Up** section. Afterwards, proceed with the class using the duplicate copy of the notebook. You can share the link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Lesson17**) with the student through the chat window.

5. When a student executes the code for the first time in the duplicate copy that you shared with them, they may get the following warning:

   ```
    Warning: This notebook was not authored by Google. This notebook was 
    authored by xyz@gmail.com. It may request access to your data stored 
    with Google such as files, emails and contacts. Please review the source 
    code and contact the creator of this notebook at xyz@gmail.com with any 
    additional questions. 

                                                      Cancel. Run Anyway.
  ```

6. Ask the student to click on the `Run Anyway` button.

7. Wherever you see the hat (**^**) sign on the heading of activity, give a hats-off to the student at the end of the activity.

    - Single hat sign, i.e., **^** denotes give hats-off for concentration. You have to assess the concentration level of a student based on their attentiveness and understanding of the concept.

    - Double hat signs, i.e., **^^** denotes give hats-off for creativity. It can be an alternative approach to writing a code for an activity or applying the logic in a totally different problem statement.

    - Triple hat signs, i.e., **^^^** denotes give hats-off for persistence. It is the ability of a student to not give up on writing code and writing code with minimal teacher support.

8. Every time you execute your code in the Google Colab notebook, please don't forget to save it by pressing the **Ctrl + S** keys (or **Command + S** keys if you are using a Mac). While saving the notebook, you or student may encounter the following error:

   ```
    Save failed
    
    The notebook has been changed outside of this session. Would you like to 
    overwrite existing changes?
    
                                                              CANCEL    YES
   ```

   Click on the `YES` button.

9. Occasionally, you may encounter some error such as `NameError, ReferenceError, ImportError` or `ModuleNotFoundError` after executing the code in $k^{th}$ code cell (where $k > 0$). Most likely, due to poor internet speed, the Colab notebook might have lost the information that all the previous codes have been executed already. As a remedy, run codes in the code cells beginning from the first code cell till the $(k - 1)^{th}$ code cell. For e.g., if the code in the $5^{th}$ code cell fails to run because of one of the aforementioned errors, execute the codes in all the first four code cells again.

10. For every **Teacher Action**, the teacher is supposed to share their screen with the student. Similarly, for every **Student Action**, the student is supposed to share their screen with the teacher.


---

### Warm-Up

```
TEACHER
  Hi student_name! How are you doing?

EXPECTED STUDENT RESPONSE
  I am doing good, teacher. How about you?


TEACHER
  I am having a good time, thank you. So, how was your day? What did you do 
  today?

EXPECTED STUDENT RESPONSE 
  I went to school. Had my lunch. Played a game and then did my homework 
  before this class.


TEACHER
  That's great. I am glad that you are having a good day. Now, let's do a 
  quick recap of our previous class. Can you recall the concepts you 
  learnt in the previous lesson?

EXPECTED STUDENT RESPONSE
  We learnt how to deploy the Random Forest Classifier prediction model 
  and how to separate the feature and target variables using the 'iloc[]'
  function.


TEACHER
  Excellent! You recalled everything that you learnt in the last class. 
  Great job! Keep it up. 
  I have an exciting quiz question for you! Are you ready to answer this question?    
  ```   
  **Instructions for the Teacher:**
  -  Please click on the "Quiz Time" button on the bottom right corner of your screen to start the In-Class Quiz.<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/quiz-time.png' width=100>   
  - A quiz will be visible to both you and the student. Encourage the student to answer the quiz question. 
  - The student may choose the wrong option, help the student to think correctly about the question and then answer again. 
  - After the student selects the correct option, the "End Quiz" button will start appearing on your screen.  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/end-quiz.png' width=100>  
  - Click the "End quiz" to close the quiz pop-up and continue the class. 
  - Do not spend more than 2 minutes on this quiz.

```
TEACHER
  Welcome to the Capstone Class. 

    Before we hop onto the Capstone challenge , let us complete today's task. 
    Our goal is to create a confusion matrix and calculate the F1-score value. 
  
  So, are you excited for this class?

EXPECTED STUDENT RESPONSE
  Yes, teacher.


TEACHER
  All right. Let's get started. I'm sharing the link to the Colab notebook 
  for this class with you. Open it up.

  Note:
    - Share the duplicate copy of the notebook that you just created with 
      the student through the chat window.
    
EXPECTED STUDENT RESPONSE
  The student opens the Colab notebook.
```

---

### Teacher-Student Activities

In the previous class, we deployed the `RandomForestClassifier` prediction model which classified the stars, in the test dataset, as `1` and `2`.

The model turned out to be 99% accurate. However, in the dataset, `565` stars are classified as `1` and remaining `5` are classified as `2`. But the Random Forest Classifier model classified every star as `1`. Ideally, it should classify a few stars as `2` because the ultimate goal of the Kepler Space telescope was to find exoplanets in space.

Hence, we need to ensure that for any kind of uneven distribution of data in the test dataset, our model should make accurate predictions. For this purpose, we need to evaluate the model that we deployed.

Generally, a classification model (in this case, Random Forest Classification) is evaluated through a concept called **confusion matrix**.

In this class, we will learn how to evaluate the performance of a classification-based machine learning model using a confusion matrix.


Let's run all the codes in the code cells that we have already covered in the previous classes and begin this class from **Activity 1: The Confusion Matrix** section. You too run the code cells until the first activity.


---

#### Loading The Datasets
Create a Pandas DataFrame every time you start the Jupyter notebook.

Dataset links (don't click on them):

1. Train dataset

  https://s3-student-datasets-bucket.whjr.online/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTrain.csv

2. Test dataset

  https://s3-student-datasets-bucket.whjr.online/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTest.csv

In [ ]:
# Load the datasets.
import pandas as pd

exo_train_df = pd.read_csv('https://s3-student-datasets-bucket.whjr.online/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTrain.csv')
exo_test_df = pd.read_csv('https://s3-student-datasets-bucket.whjr.online/whitehat-ds-datasets/kepler-exoplanets-dataset/exoTest.csv')

Check the number of rows and columns in the DataFrames.

In [ ]:
# Number of rows and columns in the DataFrames.
print(exo_train_df.shape)
exo_test_df.shape

(5087, 3198)


(570, 3198)

---

#### The `value_counts()` Function

To compute how many times a value occurs in a series, use the `value_counts()` function.

In [ ]:
# The number of times a value occurs in a Pandas series.
exo_test_df['LABEL'].value_counts()

1    565
2      5
Name: LABEL, dtype: int64

There are `565` stars which are classified as `1` and `5` stars classified as `2`. This means that only `5` stars have a planet.


---

#### Importing `RandomForestClassifier` Module
We need to import a module called `RandomForestClassifier` from a package called `sklearn.ensemble`. The `sklearn` or **scikit-learn** is a collection of many machine learning modules. Almost every machine learning algorithm can be directly applied without a knowledge of math using the **scikit-learn** library. It is kind of a plug-and-play device.

In [ ]:
# Import the 'RandomForestClassifier' module from the 'sklearn.ensemble' library.
from sklearn.ensemble import RandomForestClassifier

---

#### The Target & Feature Variables Separation

The `RandomForestClassifier` module has a function called `fit()` which takes two inputs. The first input is the collection of feature variables. The second input is the target variable. Hence, we need to extract the target variable and the feature variables separately from the training dataset.

Let's store the feature variables in the `x_train` variable and the target variable in the `y_train`.

In [ ]:
# Extract feature variables from the training dataset.
x_train = exo_train_df.iloc[:, 1:]
x_train.head()

FLUX.1   FLUX.2   FLUX.3  ...  FLUX.3195  FLUX.3196  FLUX.3197
0    93.85    83.81    20.10  ...      61.42       5.08     -39.54
1   -38.88   -33.83   -58.54  ...       6.46      16.00      19.93
2   532.64   535.92   513.73  ...     -28.91     -70.02     -96.67
3   326.52   347.39   302.35  ...     -17.31     -17.35      13.98
4 -1107.21 -1112.59 -1118.95  ...    -384.65    -411.79    -510.54

[5 rows x 3197 columns]

In [ ]:
# Retrieve only the first column, i.e., the 'LABEL' column.
y_train = exo_train_df.iloc[:, 0] 
y_train.head()

0    2
1    2
2    2
3    2
4    2
Name: LABEL, dtype: int64

---

#### Fitting The Model
Let's train the model using the `fit()` function.

In [ ]:
# Train the 'RandomForestClassifier' model using the 'fit()' function.
rf_clf = RandomForestClassifier(n_jobs=-1, n_estimators=50)
rf_clf.fit(x_train, y_train)

rf_clf.score(x_train, y_train)

1.0

As you can see, we have built the Random Forest Classifier model with 50 decision trees. The fitting accuracy score of the model is 100%.



---

#### Target And Feature Variables From Test Dataset
Now we need to make predictions on the test dataset. So, we just need to extract feature variables from the test dataset using the `iloc[]` function.

In [ ]:
# Extract the feature variables from the test dataset.
x_test = exo_test_df.iloc[:, 1:]
x_test.head()

FLUX.1   FLUX.2   FLUX.3  ...  FLUX.3195  FLUX.3196  FLUX.3197
0   119.88   100.21    86.46  ...      35.78     269.43      57.72
1  5736.59  5699.98  5717.16  ...   -2366.19   -2294.86   -2034.72
2   844.48   817.49   770.07  ...    -162.68     -36.79      30.63
3  -826.00  -827.31  -846.12  ...    -120.81    -257.56    -215.41
4   -39.57   -15.88    -9.16  ...     -61.98     -69.34     -17.84

[5 rows x 3197 columns]

Let's also extract the target variable from the test dataset so that we can compare the actual target values with the predicted values later.

In [ ]:
# Extract the target variable from the test dataset.
y_test = exo_test_df.iloc[:, 0]
y_test.head()

0    2
1    2
2    2
3    2
4    2
Name: LABEL, dtype: int64

---

#### The `predict()` Function
Now, let's make predictions on the test dataset by calling the `predict()` function with the features variables of the test dataset as an input.

In [ ]:
# Make predictions using the 'predict()' function.
y_predicted = rf_clf.predict(x_test)
y_predicted

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

The actual target values are stored in a Pandas series. So, for the sake of consistency, let's convert the NumPy array of the predicted values into a Pandas series.

In [ ]:
# Convert the NumPy array of predicted values into a Pandas series.
y_predicted = pd.Series(y_predicted)
y_predicted.head()

0    1
1    1
2    1
3    1
4    1
dtype: int64

Now, let's count the number of stars classified as `1` and `2`.

In [ ]:
# Using the 'value_counts()' function, count the number of times 1 and 2 occur in the predicted values.
y_predicted.value_counts()

1    570
dtype: int64

---

#### Activity 1: The Confusion Matrix
Let's quickly first create a confusion matrix and then will try to understand it.

To create a confusion matrix, first import `confusion_matrix` module from the `sklearn.metrics` library. This library contains all the parameters to evaluate a machine learning model. In addition to the `confusion_matrix` module, let's also import the `classification_report` module. We will use them later to evaluate our module.

In [ ]:
# Teacher Action: Import the 'confusion_matrix' and 'classification_report' functions from the 'sklearn.metrics' module.
from sklearn.metrics import confusion_matrix, classification_report

Now, create the confusion matrix using the `confusion_matrix()` function. It requires two inputs. The first input is actual target values (`y_test`) and the second input is predicted target values (`y_predicted`).

In [ ]:
# Teacher Action: Create a confusion matrix using the 'y_test' and 'y_predicted' values.
confusion_matrix(y_test, y_predicted)

array([[565,   0],
       [  5,   0]])

Now that we have got our confusion matrix, let's try to understand this concept.

**Confusion Matrix:**
---

It is way of evaluating the performance of your machine learning algorithm.

**For Example:**

Suppose that you attempted an online exam in which you already know that out of 100 questions, you have given 75 correct answers and 25 incorrect answers.

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/WHJ-BOY-TYPING-APT-C17.gif" height=400/>


However, the exam software did not assessed the answers correctly and marked many correct answers as incorrect and incorrect answers as correct. Let us evaluate the performance of this software using confusion matrix.

- There are two possible classes:
  1. Class `correct`.
  2. Class `incorrect`.

We need to find out how many correct answers were accurately assessed or predicted by the software. 

Thus, 
- positive outcome $\Rightarrow$ `correct` answer.
- negative outcome $\Rightarrow$ `incorrect` answer.

In technical terms, the desired outcome is called a **positive outcome**. 


Now, consider the following table and a `2 X 2` matrix known as **confusion matrix**. This table shows the actual and predicted values for the first 4 questions.

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/table2.PNG"/>

```
TEACHER
   here the first row indicates that the answer to the first question was 
   actually incorrect and the system also marked it incorrect. Thus, the 
   first question was accurately assessed by the software.

   Similarly, the third row indicates that the answer to the third question was 
   actually correct, however the system marked it as incorrect. Thus the 
   software performed incorrect assessment of the third question
```

Now let us have a look at each cell of the confusion matrix.

1. The first row first column value indicates those `incorrect` answers which were <b><font color=green>accurately</font></b> assessed or predicted as `incorrect` by the software.
Such values are called as **True Negative (TN)**.


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/True_negative.png"/>

2. The second  row second column value  indicates those `correct` answers which were <b><font color=green>accurately</font></b> assessed or predicted as `correct` by the software.
Such values are called as **True Positive (TP)**.

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/true_positives.png"/>

3. The second  row first column value  indicates those `correct`  answers which were <b><font color=red>inaccurately</font></b>  assessed or predicted as `incorrect` by the software.
Such values are called as **False Negative (FN)**.


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/fn.png"/>

4. The first  row second column value  indicates those `incorrect`  answers which were <b><font color=red>inaccurately</font></b>  assessed or predicted as `correct` by the software.
Such values are called **False Positive (FP)**.


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/FP.png"/>


The resultant confusion matrix obtained after evaluating values for all the 100 questions are as follows:

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-negative-positive-apt-c17-01.png" height=300/>


- Values that are accurately predicted or assessed by the model are labelled **True (T)**. Thus, the number of answers which were <b><font color=green>accurately</font></b> predicted by the software = `85`


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-85-ans-apt-c17.gif" height=450/>

- Values that are inaccurately predicted or assessed by the model are labelled **False (F)**. Thus, the number of answers which were <b><font color=red>inaccurately</font></b> predicted by the software = `15`


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-15-answers-apt-c17.gif" height=450/>

Thus the confusion matrix compares the actual values with the predicted values and thus it is very useful in evaluating the performance of your machine learning model.

---

Let us apply the concept of confusion matrix for our dataset.

There are two possible classes:
1. The class `1` values are stars **NOT** having a planet.
2. The class `2` values are stars having a planet.

 So, after you deploy the classification model, there are 4 possible outcomes. They are:

1. Class `1` values predicted as class `1`. 

2. Class `1` values predicted as class `2`.

3. Class `2` values predicted as class `1`.

4. Class `2` values predicted as class `2`.

These 4 possibilities can be reported in a confusion matrix. 

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|||
|Actual Class `2` (`y_test`)|||

where

- `y_test` contains the actual class `1` and class `2` values

- `y_predicted` contains the predicted class `1` and class `2` values

In this table,

- the values **predicted as class `1` and actually belonging to class `1`** are reported in the first row and first column.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565||
|Actual Class `2` (`y_test`)|||

- the values **predicted as class `1` but actually belonging to class `2`** are reported in the second row and first column. 

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565||
|Actual Class `2` (`y_test`)|5||

- the values **predicted as class `2` and actually belonging to class `2`** are reported in the second row and second column.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565||
|Actual Class `2` (`y_test`)|5|0|

- the values **predicted as class `2` but actually belonging to class `1`** are reported in the first row and second column.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565|0|
|Actual Class `2` (`y_test`)|5|0|

In this case, the class `1` values refer to the stars not having a planet whereas class `2` values refer to the stars having a planet. 

**Positive Outcome**

Detecting a star having a planet is the desired outcome (positive outcome). 
Thus, 
- positive outcome $\Rightarrow$ class `2`.

- negative outcome $\Rightarrow$ class `1`.

So, here the positive outcome is the prediction of the stars having a planet, i.e., prediction of the class `2` values. Likewise, finding a star which does not have any planet is a *negative outcome*. So, here the negative outcome is the prediction of the class `1` values. 

Observe the output of `confusion_matrix(y_test, y_predicted)` function.

```
array([[565,   0],
       [  5,   0]])
```




- `565` values are **True Negative (TN)** values because they are **truly** predicted as class `1` values.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565 (TN)||
|Actual Class `2` (`y_test`)|||


- `5` values are **False Negative (FN)** values because they are **falsely** predicted as class `1` values. They should have been predicted as class `2` values.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565 (TN)||
|Actual Class `2` (`y_test`)|5 (FN)||


- `0` values are **True Positive (TP)** values because they are **truly** predicted as class `2` values.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565 (TN)||
|Actual Class `2` (`y_test`)|5 (FN)|0 (TP)|


- `0` values are **False Positive (FP)** values because they are **falsely** predicted as class `2` values. They should have been predicted as class `1` values.

||Predicted Class `1` (`y_predicted`)|Predicted Class `2` (`y_predicted`)|
|-|-|-|
|Actual Class `1` (`y_test`)|565 (TN)|0 (FP)|
|Actual Class `2` (`y_test`)|5 (FN)|0 (TP)|

---

#### Activity 2: Precision And Recall^

A good prediction model provides a very large number of true positive (TP) values and a very low number of true negative (TN) values.

**Precision:**

Based on the TP and FP values, we define a parameter called **precision**. 
It is used to evaluate the number of correct positive predictions made.

It is the ratio of the TP values to the sum of TP and FP values, i.e.,

(defn of precision and recall)
(precision of both outcomes.)
(add example values)

$$\text{precision} = \frac{\text{TP}}{\text{TP + FP}}$$

Let us calculate the precision for exam software:

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-negative-positive-apt-c17-01.png" height=300/>

For the above matrix,
$$\text{precision} = \frac{\text{TP}}{\text{TP + FP}}=\frac{\text{65}}{\text{65 + 5}}=\text{0.928}$$

Ideally, the precision should be 1 for a good classifier model. In this case, it is 0.928 which is quite good.

Similarly, let's calculate the precision for our Random Forest classifier model.
Currently, the model has given `0` TP values and `0` FP values. Therefore, the precision value is undefined because

$$\text{precision} = \frac{0}{0 + 0} = \text{undefined}$$

*In mathematics, the division by 0 is undefined (or not defined).*

**Recall:**

Based on the TP and FN values, we define another parameter called **recall**.
It is the ratio of the TP values to the sum of TP and FN values, i.e, 

$$\text{recall} = \frac{\text{TP}}{\text{TP + FN}}$$

Let us calculate the precision for the exam software model:
<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-negative-positive-apt-c17-01.png" height=300/>

For the above matrix,
$$\text{recall} = \frac{\text{TP}}{\text{TP + FN}}=\frac{\text{65}}{\text{65 + 10}}=\text{0.867}$$

Ideally, the recall should be 1 for a good classifier model. In this case, it is 0.867 which is pretty good.

Similarly, let's calculate the recall for our Random Forest classifier model.

Currently, the model gives `0` TP and `5` FN values. Hence, the recall value is 0 because

$$\text{recall} = \frac{0}{0+5} = \text{0}$$

Imagine if the prediction model labels every star as `2`, i.e, every star has a planet. Then, the number of TP values will be the maximum, i.e., `5` but the number of FP values will also be maximum, i.e., `565`. In such a case, the precision value would be

$$\text{precision} = \frac{5}{5+565} = \frac{5}{570} = 0.008$$

which is very very low.

Also, the model will give `0` FN values. Then, the recall value would be

$$\text{recall} = \frac{5}{5 + 0} = 1$$


So, even though the recall value would be equal to 1, the precision value would be close to 0. Hence, this would be a bad prediction model.


Evidently, there is a trade-off. If the recall value is high, then the precision value will be low and vice-versa. Hence, we need to find an optimum point where both, the precision and the recall values are acceptable.

---

#### Activity 3: The `f1-score`

To find an optimum point where both, the precision and recall values, are high, we calculate another parameter called **f1-score**. It is a harmonic mean of the precision and recall values, i.e.,



$$\text{f1-score} = 2 \left( \frac{\text{precision} \times \text{recall}}{\text{precision} + \text{recall}} \right)$$

Let us calculate the f1-score for the exam software model:
<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C17/whj-negative-positive-apt-c17-01.png" height=300/>

For the above matrix,
$$\text{f1-score} = 2 \left( \frac{\text{0.928} \times \text{0.867}}{\text{0.928} + \text{0.867}} \right)=\text{0.896}$$

f1-score will be high only when both precision and recall are high. In this case, it is 0.896 which is a good f1-score.

Similarly, let's calculate the f1-score for our Random Forest classifier model.

Based on the current predictions, the f1-scores value is undefined because both the precision and recall values are also undefined.

$$\text{f1-score} = 2 \left( \frac{\text{undefined} \times 0}{\text{undefined} + 0} \right) =  \text{undefined}$$

You can also get these values by calling a function called `classification_report()`. It takes two inputs: the actual target values and the predicted target values, i.e., `y_test` and `y_predicted`.

**Note:** You may get the following warning message after executing the code in the code cell below.

```
Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
```

Ignore the warning.

In [ ]:
# Student Action: Print the 'precision', 'recall' and 'f1-score' values using the 'classification_report()' function.
print(classification_report(y_test, y_predicted))

              precision    recall  f1-score   support

           1       0.99      0.99      0.99       565
           2       0.00      0.00      0.00         5

    accuracy                           0.98       570
   macro avg       0.50      0.50      0.50       570
weighted avg       0.98      0.98      0.98       570



As you can see, the `precision` and `f1-score` values are reported as `0.00` for class `2` because they are actually undefined values. 
  
Ideally, the above values for class `2` should also be close to `1.00`. Then only we can say that our prediction model is satisfactory. This shows that accuracy alone cannot tell whether a prediction model is making correct predictions or not.

In the next class, we will try to improve the model so that we get the desired precision, recall and f1-score values for class `2`.



---

### Additional Activities

The activities starting from this point are optional. Please do these activities **ONLY** if you have time to spare in the class. Otherwise, skip to the **Wrap-Up** section. The additional activities will not be available in the class copy of the notebook. You will have to manually add these activities in the class copy by adding new text and code cells.

Moreover, you don't have to do all the additional activities. Depending on the availability of time in a class, you can choose the number of additional activities to perform from this collection. 

---

#### Activity 1: Google Coding Challenge

This coding question is taken from Google Coding Challenge

**Problem**

There are $N$ houses for sale. The $i^{\text{th}}$ house costs $A_i$ dollars to buy. You have a budget of $B$ dollars to spend. What is the maximum number of houses you can buy?

**Input**

The first line of the input gives the number of test cases, $T$. $T$ test cases follow. Each test case begins with a single line containing the two integers $N$ and $B$. The second line contains $N$ integers. The $i^{\text{th}}$ integer is $A_i$, the cost of the $i^{\text{th}}$ house.

**Output**

For each test case, output one line containing `Case #x: y`, where `x` is the test case number (starting from 1) and `y` is the maximum number of houses you can buy.

**Limits**

$1 ≤ T ≤ 100$$

$1 ≤ B ≤ 10^5$

$1 ≤ A_i ≤ 1000$, for all $i$.


**Sample** 

**Input** 

```
3
4 100
20 90 40 90
4 50
30 30 10 10
3 300
999 999 999
```

**Output**

```
Case # 1 : 2
Case # 2 : 3
Case # 3 : 0
```

In Sample `Case #1`, you have a budget of `100` dollars. You can buy the first and third houses for `20 + 40 = 60` dollars.

In Sample `Case #2`, you have a budget of `50` dollars. You can buy the first, third and fourth houses for `30 + 10 + 10 = 50` dollars.

In Sample `Case #3`, you have a budget of `300` dollars. You cannot buy any houses. So the answer is `0`.

In [ ]:
# Solution:
def max_affordable_houses(my_budget, house_prices):
	sum_house_prices = 0
	count = 0
	house_prices.sort()
	for price in house_prices:
		sum_house_prices += int(price)
		if sum_house_prices > my_budget:
			continue
		else:
			count += 1
	return count

test_cases = [] # This line of code should be shown to the students.
num_test_cases = int(input()) # This line of code should be shown to the students.
while num_test_cases > 0: # This line of code should be shown to the students.
	input_list = input().split(' ') # This line of code should be shown to the students.
	num_houses = int(input_list[0]) # This line of code should be shown to the students.
	my_budget = int(input_list[1]) # This line of code should be shown to the students.
	house_prices = input().split(' ') # This line of code should be shown to the students.
	house_prices = [int(price) for price in house_prices] # This line of code should be shown to the students.
	test_cases.append([my_budget, house_prices]) # This line of code should be shown to the students.
	num_test_cases -= 1 # This line of code should be shown to the students.

for i in range(len(test_cases)):
	my_budget = test_cases[i][0]
	house_prices = test_cases[i][1]
	print("Case #", i + 1, ":", max_affordable_houses(my_budget, house_prices))

3
4 100
20 90 40 90
4 50
30 30 10 10
3 100
999 999 999
Case # 1 : 2
Case # 2 : 3
Case # 3 : 0


---

### Wrap-Up

```
TEACHER
  Did you enjoy the class?

EXPECTED STUDENT RESPONSE
  Yes, teacher.


TEACHER
  What part of the class did you enjoy?

EXPECTED STUDENT RESPONSE
  I enjoyed everything.


TEACHER
  I am glad you are having a good time. But make sure that you practice 
  writing code in your free time.

EXPECTED STUDENT RESPONSE
  Sure teacher.


TEACHER
  Congratulations! You've reached a milestone. You can now attempt Project 
  Gender Voices Classification. In this project, you will be provided with a 
  dataset which contains some statistical information about the audio 
  frequencies of different male and female voices. Based on the information 
  provided, you have to find out which voice belongs to which gender using the 
  Random Forest Classifier algorithm.
  
  In the next class, we will learn how to process the training dataset so 
  that our prediction model makes accurate predictions. This step is 
  called data processing. It enables a prediction model (or algorithm) 
  to learn the properties of data in an effective manner so that it can 
  distinguish between different types of data in a dataset.
  
EXPECTED STUDENT RESPONSE
  Sure teacher.


TEACHER
  See you in the next class. Bye!

EXPECTED STUDENT RESPONSE
  Bye teacher!
```

---

### Activities

**Teacher Activities**

1. Hunting Exoplanets In Space - Model Evaluation (Class Copy)

   https://colab.research.google.com/drive/1dsrP-6aEDKa05xhfq6ewWYG2bQtfBS1j

2. Hunting Exoplanets In Space - Model Evaluation (Reference)

   https://colab.research.google.com/drive/1H1SDf0Iuqf-BkZXSs9YXXyJvlnssMFRG

---